File: frontend/src/services/userService.ts
// 用户服务，处理用户认证、收藏和浏览历史相关的API调用
import { API_BASE_URL } from '../config/api';

// 定义响应类型
export interface ApiResponse<T> {
  data?: T;
  error?: {
    code: string;
    message: string;
  };
  success: boolean;
}

// 用户信息
export interface UserInfo {
  id: number;
  username: string;
  email: string;
  avatar: string | null;
  created_at?: string;
}

// 用户注册请求
export interface RegisterRequest {
  username: string;
  email: string;
  password: string;
  avatar?: string;
}

// 用户登录请求
export interface LoginRequest {
  username: string;
  password: string;
}

// 收藏/历史操作响应
export interface MessageResponse {
  message: string;
}

// 检查收藏状态响应
export interface CheckFavoriteResponse {
  is_favorite: boolean;
}

// 新闻摘要信息
export interface NewsSummary {
  id: number;
  title: string;
  summary: string;
  publish_time: string;
  source: string;
}

// 收藏项
export interface FavoriteItem {
  id: number;
  news_id: number;
  created_at: string;
  news: NewsSummary;
}

// 历史记录项
export interface HistoryItem {
  id: number;
  news_id: number;
  viewed_at: string;
  news: NewsSummary;
}

// 分页响应
export interface PaginatedResponse<T> {
  total: number;
  page: number;
  page_size: number;
  items: T[];
}

// 会话信息
export interface UserSession {
  id: string;
  user_id: number;
  title: string;
  created_at: string;
  updated_at: string;
}

// 会话列表响应
export interface SessionsResponse {
  sessions: UserSession[];
}

// 创建会话请求
export interface CreateSessionRequest {
  title: string;
}

// 创建会话响应
export interface CreateSessionResponse {
  session_id: string;
}



/**
 * 用户注册
 * @param userData 用户注册数据
 * @returns API响应
 */
export async function register(userData: RegisterRequest): Promise<ApiResponse<UserInfo>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/user/register`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(userData),
    });
    
    return await response.json();
  } catch (error) {
    console.error('注册失败:', error);
    return {
      success: false,
      error: {
        code: 'NETWORK_ERROR',
        message: '网络错误，请稍后重试',
      },
    };
  }
}

/**
 * 用户登录
 * @param credentials 登录凭证
 * @returns API响应
 */
export async function login(credentials: LoginRequest): Promise<ApiResponse<UserInfo>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/user/login`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(credentials),
    });
    
    return await response.json();
  } catch (error) {
    console.error('登录失败:', error);
    return {
      success: false,
      error: {
        code: 'NETWORK_ERROR',
        message: '网络错误，请稍后重试',
      },
    };
  }
}

/**
 * 添加收藏
 * @param userId 用户ID
 * @param newsId 新闻ID
 * @returns API响应
 */
export async function addFavorite(userId: number, newsId: number): Promise<ApiResponse<MessageResponse>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/user/${userId}/favorites`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ news_id: newsId }),
    });
    
    return await response.json();
  } catch (error) {
    console.error('添加收藏失败:', error);
    return {
      success: false,
      error: {
        code: 'NETWORK_ERROR',
        message: '网络错误，请稍后重试',
      },
    };
  }
}

/**
 * 取消收藏
 * @param userId 用户ID
 * @param newsId 新闻ID
 * @returns API响应
 */
export async function removeFavorite(userId: number, newsId: number): Promise<ApiResponse<MessageResponse>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/user/${userId}/favorites/${newsId}`, {
      method: 'DELETE',
    });
    
    return await response.json();
  } catch (error) {
    console.error('取消收藏失败:', error);
    return {
      success: false,
      error: {
        code: 'NETWORK_ERROR',
        message: '网络错误，请稍后重试',
      },
    };
  }
}

/**
 * 获取收藏列表
 * @param userId 用户ID
 * @param page 页码
 * @param pageSize 每页数量
 * @returns API响应
 */
export async function getFavorites(userId: number, page: number = 1, pageSize: number = 20): Promise<ApiResponse<PaginatedResponse<FavoriteItem>>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/user/${userId}/favorites?page=${page}&page_size=${pageSize}`);
    
    return await response.json();
  } catch (error) {
    console.error('获取收藏列表失败:', error);
    return {
      success: false,
      error: {
        code: 'NETWORK_ERROR',
        message: '网络错误，请稍后重试',
      },
    };
  }
}

/**
 * 检查是否收藏
 * @param userId 用户ID
 * @param newsId 新闻ID
 * @returns API响应
 */
export async function checkFavorite(userId: number, newsId: number): Promise<ApiResponse<CheckFavoriteResponse>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/user/${userId}/favorites/${newsId}/check`);
    
    return await response.json();
  } catch (error) {
    console.error('检查收藏状态失败:', error);
    return {
      success: false,
      error: {
        code: 'NETWORK_ERROR',
        message: '网络错误，请稍后重试',
      },
    };
  }
}

/**
 * 添加浏览历史
 * @param userId 用户ID
 * @param newsId 新闻ID
 * @returns API响应
 */
export async function addHistory(userId: number, newsId: number): Promise<ApiResponse<MessageResponse>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/user/${userId}/history`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ news_id: newsId }),
    });
    
    return await response.json();
  } catch (error) {
    console.error('添加浏览历史失败:', error);
    return {
      success: false,
      error: {
        code: 'NETWORK_ERROR',
        message: '网络错误，请稍后重试',
      },
    };
  }
}

/**
 * 获取浏览历史
 * @param userId 用户ID
 * @param page 页码
 * @param pageSize 每页数量
 * @returns API响应
 */
export async function getHistory(userId: number, page: number = 1, pageSize: number = 20): Promise<ApiResponse<PaginatedResponse<HistoryItem>>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/user/${userId}/history?page=${page}&page_size=${pageSize}`);
    
    return await response.json();
  } catch (error) {
    console.error('获取浏览历史失败:', error);
    return {
      success: false,
      error: {
        code: 'NETWORK_ERROR',
        message: '网络错误，请稍后重试',
      },
    };
  }
}

/**
 * 获取用户会话列表
 * @param userId 用户ID
 * @returns API响应
 */
export async function getUserSessions(userId: number): Promise<ApiResponse<SessionsResponse>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/user/${userId}/sessions`);
    
    return await response.json();
  } catch (error) {
    console.error('获取用户会话列表失败:', error);
    return {
      success: false,
      error: {
        code: 'NETWORK_ERROR',
        message: '网络错误，请稍后重试',
      },
    };
  }
}

/**
 * 创建用户会话
 * @param userId 用户ID
 * @param sessionData 会话数据
 * @returns API响应
 */
export async function createUserSession(userId: number, sessionData: CreateSessionRequest): Promise<ApiResponse<CreateSessionResponse>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/user/${userId}/session/create`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(sessionData),
    });
    
    return await response.json();
  } catch (error) {
    console.error('创建用户会话失败:', error);
    return {
      success: false,
      error: {
        code: 'NETWORK_ERROR',
        message: '网络错误，请稍后重试',
      },
    };
  }
}

File: backend/services/user_service.py
from sqlalchemy.orm import Session
from sqlalchemy import desc, or_
from spiders.database_news import User, UserNewsFavorites, UserNewsHistory, News, Session
from typing import List, Optional, Dict, Any
from datetime import datetime
import hashlib
from extensions import bcrypt

class UserService:
    def __init__(self, db: Session):
        self.db = db
    
    def register_user(self, username: str, email: str, password: str, avatar: Optional[str] = None) -> Optional[Dict[str, Any]]:
        """
        用户注册
        
        Args:
            username: 用户名
            email: 邮箱
            password: 密码
            avatar: 头像URL（可选）
            
        Returns:
            用户信息字典，如果注册失败则返回None
        """
        # 检查用户名是否已存在
        existing_user = self.db.query(User).filter(
            or_(User.username == username, User.email == email)
        ).first()
        
        if existing_user:
            return None
        
        # 密码哈希
        password_hash = bcrypt.generate_password_hash(password).decode('utf-8')
        
        # 创建用户
        user = User(
            username=username,
            email=email,
            password_hash=password_hash,
            avatar=avatar
        )
        
        self.db.add(user)
        self.db.commit()
        self.db.refresh(user)
        
        return {
            "id": user.id,
            "username": user.username,
            "email": user.email,
            "avatar": user.avatar,
            "created_at": user.created_at.isoformat() if user.created_at else None
        }
    
    def authenticate_user(self, username: str, password: str) -> Optional[Dict[str, Any]]:
        """
        用户认证
        
        Args:
            username: 用户名或邮箱
            password: 密码
            
        Returns:
            用户信息字典，如果认证失败则返回None
        """
        # 查找用户
        user = self.db.query(User).filter(
            or_(User.username == username, User.email == username)
        ).first()
        
        if not user:
            return None
        
        # 验证密码
        if not bcrypt.check_password_hash(user.password_hash, password):
            return None
        
        return {
            "id": user.id,
            "username": user.username,
            "email": user.email,
            "avatar": user.avatar,
            "created_at": user.created_at.isoformat() if user.created_at else None
        }
    

    def add_favorite(self, user_id: int, news_id: int) -> bool:
        """
        添加收藏
        
        Args:
            user_id: 用户ID
            news_id: 新闻ID
            
        Returns:
            是否成功添加
        """
        # 检查是否已收藏
        existing = self.db.query(UserNewsFavorites).filter(
            UserNewsFavorites.user_id == user_id,
            UserNewsFavorites.news_id == news_id
        ).first()
        
        if existing:
            return False
        
        # 添加收藏
        favorite = UserNewsFavorites(user_id=user_id, news_id=news_id)
        self.db.add(favorite)
        self.db.commit()
        
        return True
    
    def remove_favorite(self, user_id: int, news_id: int) -> bool:
        """
        取消收藏
        
        Args:
            user_id: 用户ID
            news_id: 新闻ID
            
        Returns:
            是否成功取消
        """
        favorite = self.db.query(UserNewsFavorites).filter(
            UserNewsFavorites.user_id == user_id,
            UserNewsFavorites.news_id == news_id
        ).first()
        
        if not favorite:
            return False
        
        self.db.delete(favorite)
        self.db.commit()
        
        return True
    
    def get_favorites(self, user_id: int, page: int = 1, page_size: int = 20) -> Dict[str, Any]:
        """
        获取用户收藏列表
        
        Args:
            user_id: 用户ID
            page: 页码
            page_size: 每页数量
            
        Returns:
            收藏列表和分页信息
        """
        query = self.db.query(UserNewsFavorites).filter(
            UserNewsFavorites.user_id == user_id
        ).order_by(desc(UserNewsFavorites.created_at))
        
        total = query.count()
        total_pages = (total + page_size - 1) // page_size
        
        favorites = query.offset((page - 1) * page_size).limit(page_size).all()
        
        # 获取新闻详情
        news_list = []
        for fav in favorites:
            if fav.news:
                news_list.append({
                    "id": fav.news.id,
                    "title": fav.news.title,
                    "content": fav.news.content,
                    "source": fav.news.source,
                    "publish_time": fav.news.publish_time.isoformat() if fav.news.publish_time else None,
                    "views": fav.news.views,
                    "has_image": fav.news.has_image,
                    "image_url": fav.news.image_url,
                    "summary": fav.news.summary,
                    "favorited_at": fav.created_at.isoformat() if fav.created_at else None
                })
        
        return {
            "news": news_list,
            "pagination": {
                "page": page,
                "page_size": page_size,
                "total": total,
                "total_pages": total_pages
            }
        }
    
    def is_favorite(self, user_id: int, news_id: int) -> bool:
        """
        检查是否已收藏
        
        Args:
            user_id: 用户ID
            news_id: 新闻ID
            
        Returns:
            是否已收藏
        """
        favorite = self.db.query(UserNewsFavorites).filter(
            UserNewsFavorites.user_id == user_id,
            UserNewsFavorites.news_id == news_id
        ).first()
        
        return favorite is not None
    
    def add_history(self, user_id: int, news_id: int) -> bool:
        """
        添加浏览历史
        
        Args:
            user_id: 用户ID
            news_id: 新闻ID
            
        Returns:
            是否成功添加
        """
        # 添加浏览历史
        history = UserNewsHistory(user_id=user_id, news_id=news_id)
        self.db.add(history)
        self.db.commit()
        
        return True
    
    def get_history(self, user_id: int, page: int = 1, page_size: int = 20) -> Dict[str, Any]:
        """
        获取用户浏览历史
        
        Args:
            user_id: 用户ID
            page: 页码
            page_size: 每页数量
            
        Returns:
            浏览历史列表和分页信息
        """
        query = self.db.query(UserNewsHistory).filter(
            UserNewsHistory.user_id == user_id
        ).order_by(desc(UserNewsHistory.viewed_at))
        
        total = query.count()
        total_pages = (total + page_size - 1) // page_size
        
        history_list = query.offset((page - 1) * page_size).limit(page_size).all()
        
        # 获取新闻详情
        news_list = []
        for hist in history_list:
            if hist.news:
                news_list.append({
                    "id": hist.news.id,
                    "title": hist.news.title,
                    "content": hist.news.content,
                    "source": hist.news.source,
                    "publish_time": hist.news.publish_time.isoformat() if hist.news.publish_time else None,
                    "views": hist.news.views,
                    "has_image": hist.news.has_image,
                    "image_url": hist.news.image_url,
                    "summary": hist.news.summary,
                    "viewed_at": hist.viewed_at.isoformat() if hist.viewed_at else None
                })
        
        return {
            "news": news_list,
            "pagination": {
                "page": page,
                "page_size": page_size,
                "total": total,
                "total_pages": total_pages
            }
        }
    
    def get_user_sessions(self, user_id: int) -> List[Dict[str, Any]]:
        """
        获取用户的会话列表
        
        Args:
            user_id: 用户ID
            
        Returns:
            会话列表
        """
        sessions = self.db.query(Session).filter(
            Session.user_id == user_id
        ).order_by(desc(Session.updated_at)).all()
        
        return [
            {
                "session_id": session.id,
                "title": session.title,
                "created_at": session.created_at.isoformat() if session.created_at else None,
                "updated_at": session.updated_at.isoformat() if session.updated_at else None
            }
            for session in sessions
        ]
